### PR TITLE
R.info & R.entitlements

### PR DESCRIPTION
**Pull request by tomlokhorst**
Generates `R.info`, implements https://github.com/mac-cain13/R.swift/issues/414

This loops thru the Info.plist dictionary and generates static references to most keys and values.

TODO:
- [x] Store key for dictionary
- [x] Add `R.entitlements`, since that may share a lot of code with this
- [x] Arrays of dictionaries (e.g. `CFBundleURLTypes` or `CFBundleDocumentTypes`) aren't supported (support for `UIApplicationShortcutItems` is hardcoded)
- [x] Support `$(SOME_VARIABLE)` from .xcconfig files
- [ ] Support localised strings in `InfoPlist.strings`
- [ ] Support for different Info.plist files for different build configurations

### CODE REVIEW

**Comment by tomlokhorst** [comment]
This currently generates a R.info struct for all values in the info.plist file. This might be a bit overkill, but at least everything you need is in there.

This is what's currently generated for the info.plist file in ResourceApp:

<details>
<summary>ResourceApp R.info</summary>

```swift

  /// This `R.info` struct is generated, and contains static references to 22 properties.
  struct info {
    static let cfBundleDevelopmentRegion = "en"
    static let cfBundleExecutable = "$(EXECUTABLE_NAME)"
    static let cfBundleIdentifier = "$(PRODUCT_BUNDLE_IDENTIFIER)"
    static let cfBundleInfoDictionaryVersion = "6.0"
    static let cfBundleName = "$(PRODUCT_NAME)"
    static let cfBundlePackageType = "APPL"
    static let cfBundleShortVersionString = "1.0"
    static let cfBundleSignature = "????"
    static let cfBundleVersion = "1"
    static let lsRequiresIPhoneOS = true
    static let uiApplicationSupportsMultipleScenes = true
    static let uiLaunchStoryboardName = "LaunchScreen"
    static let uiMainStoryboardFile = "Main"
    
    struct cfBundleDocumentTypes {
      struct mkDirectionsRequest {
        static let _key = "MKDirectionsRequest"
        static let cfBundleTypeName = "MKDirectionsRequest"
        
        struct lsItemContentTypes {
          static let comAppleMapsDirectionsrequest = "com.apple.maps.directionsrequest"
          
          fileprivate init() {}
        }
        
        fileprivate init() {}
      }
      
      fileprivate init() {}
    }
    
    struct nsUserActivityTypes {
      static let planTripIntent = "PlanTripIntent"
      static let showDeparturesIntent = "ShowDeparturesIntent"
      
      fileprivate init() {}
    }
    
    struct uiAppFonts {
      static let averiaLibreBITtf = "AveriaLibre-BI.ttf"
      static let averiaLibreBTtf = "AveriaLibre-B.ttf"
      static let averiaLibreLTtf = "AveriaLibre-L.ttf"
      static let averiaLibreTtf = "AveriaLibre.ttf"
      static let gdyBkltter1911Ttf = "GdyBkltter1911.ttf"
      
      fileprivate init() {}
    }
    
    struct uiApplicationShortcutItems {
      struct nlMathijskadijkShortcutsQrScanning {
        static let _key = "nl.mathijskadijk.shortcuts.qr-scanning"
        static let uiApplicationShortcutItemIconFile = "ShortcutQrScanning"
        static let uiApplicationShortcutItemTitle = "Scan QR-code"
        static let uiApplicationShortcutItemType = "nl.mathijskadijk.shortcuts.qr-scanning"
        
        fileprivate init() {}
      }
      
      struct nlMathijskadijkShortcutsSendParcel {
        static let _key = "nl.mathijskadijk.shortcuts.send-parcel"
        static let uiApplicationShortcutItemIconFile = "ShortcutSendParcel"
        static let uiApplicationShortcutItemTitle = "Send a Parcel"
        static let uiApplicationShortcutItemType = "nl.mathijskadijk.shortcuts.send-parcel"
        
        fileprivate init() {}
      }
      
      fileprivate init() {}
    }
    
    struct uiRequiredDeviceCapabilities {
      static let armv7 = "armv7"
      
      fileprivate init() {}
    }
    
    struct uiSceneConfigurations {
      static let _key = "UISceneConfigurations"
      
      struct uiWindowSceneSessionRoleApplication {
        struct defaultConfiguration {
          static let _key = "Default Configuration"
          static let uiLaunchStoryboardName = "LaunchScreen"
          static let uiSceneConfigurationName = "Default Configuration"
          static let uiSceneDelegateClassName = "$(PRODUCT_MODULE_NAME).SceneDelegate"
          
          fileprivate init() {}
        }
        
        fileprivate init() {}
      }
      
      fileprivate init() {}
    }
    
    struct uiStatusBarTintParameters {
      static let _key = "UIStatusBarTintParameters"
      
      struct uiNavigationBar {
        static let _key = "UINavigationBar"
        static let style = "UIBarStyleDefault"
        static let translucent = false
        
        fileprivate init() {}
      }
      
      fileprivate init() {}
    }
    
    struct uiSupportedInterfaceOrientations {
      static let uiInterfaceOrientationLandscapeLeft = "UIInterfaceOrientationLandscapeLeft"
      static let uiInterfaceOrientationLandscapeRight = "UIInterfaceOrientationLandscapeRight"
      static let uiInterfaceOrientationPortrait = "UIInterfaceOrientationPortrait"
      
      fileprivate init() {}
    }
    
    struct uiSupportedInterfaceOrientationsIpad {
      static let uiInterfaceOrientationLandscapeLeft = "UIInterfaceOrientationLandscapeLeft"
      static let uiInterfaceOrientationLandscapeRight = "UIInterfaceOrientationLandscapeRight"
      static let uiInterfaceOrientationPortrait = "UIInterfaceOrientationPortrait"
      static let uiInterfaceOrientationPortraitUpsideDown = "UIInterfaceOrientationPortraitUpsideDown"
      
      fileprivate init() {}
    }
    
    fileprivate init() {}
  }
```

</details>

Should we add a whitelist to only support certain keys, or just include everything? What do you think @mac-cain13?

**Comment by mac-cain13** [comment]
I think including everything is okay.

If I understand your `ResourceApp R.info` snippet correctly the values in the `Info.plist` are now also placed in the `R.generated.swift` file. Shouldn't we fetch those values at runtime? Like this:

`Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String`

**Comment by tomlokhorst** [comment]
Yes, it's more useful to do that, I can't think of a reason for accessing the raw (unresolved) value from the Info.plist file.

Thinking some more about this: I do think it makes sense to add a whitelist for all the keys we know are useful. That way the struct won't be cluttered up with stuff you shouldn't use and you can easily find the strings you do need to statically refer to. If we forget something, adding something to that whitelist is easy enough.

This would be my first whitelist:
`["UIApplicationShortcutItems", "UISceneConfigurations", "NSUserActivityTypes", "NSExtension"]`

**Comment by mac-cain13** [comment]
Agreed, we can just extend the whitelist over time. 👍

**Comment by tomlokhorst** [comment]
Also, I think the open checkboxes are just missing features. They're fine to add later, when/if someone needs them. 

I propose merging this PR now and creating an issue for the missing feature.

File: Sources/RswiftCore/Generators/PropertyListGenerator.swift [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    self.plists = plists
+    self.toplevelKeysWhitelist = toplevelKeysWhitelist
+  }
+
+  func generatedStruct(at externalAccessLevel: AccessLevel, prefix: SwiftIdentifier) -> Struct {
+    guard let plist = plists.first else { return .empty }
+
+    guard plists.all(where: { $0.url == plist.url }) else {
+      let configs = plists.map { $0.buildConfigurationName }
+      warn("Build configrurations \(configs) use different \(name) files, this is not yet supported")
```
Comment (mac-cain13) [inline Sources/RswiftCore/Generators/PropertyListGenerator.swift]: Typo: `configrurations` -> `configurations`

**Comment by tomlokhorst** [comment]
Oh right, I didn't think that through...

The values are statically generated, not resolved at runtime. But that's way better, that fixes the $VARIABLE issue.
I'll fix the code generator.

**Comment by tomlokhorst** [comment]
Now loading Info.plist values at runtime from the bundle, instead of generating strings with variables in them.

**Pull request merged**